lib/daru/vector.rb

```diff
@@ -225,25 +262,43 @@ def [](*input_indexes)
     #   #  a 999
     #   #  b   2
     #   #  c   3
-    def []=(*location, value)
-      cast(dtype: :array) if value.nil? && dtype != :array
+    def []=(*indexes, val)
+      cast(dtype: :array) if val.nil? && dtype != :array
+
+      guard_type_check(val)
+
+      modify(indexes, val)
+
+      update_internal_state
+    end
+
+    def modify(indexes, val)
+      positions = @index.pos(*indexes)
 
-      guard_type_check(value)
+      if positions.is_a? Numeric
+        @data[positions] = val
+      else
+        positions.each { |pos| @data[pos] = val }
+      end
+    end
+
+    def insert(indexes, val)
+      new_index = @index.add(*indexes)
+      # May be create +=
+      (new_index.size - @index.size).times { @data << val }
+      @index = new_index
+    end
 
-      pos = @index[*location]
+    # Works similar to #[]= but also insert the vector in case index is not valid
+    # It is there only to be accessed by Daru::DataFrame and not meant for user.
+    def set indexes, val
+      cast(dtype: :array) if val.nil? && dtype != :array
```

Comment:
> Or keep it private and use a hack to access the internal data structures. Users should be kept away from expanding vectors.

This is a function to be used by Daru::DataFrame to expand itself, since vector shouldn't be able to expand itself, `set` is the function I have created. I am not sure how to prevent it from the user and make it usable only by Daru::DataFrame.

ges be broadcast to the rest?